File: trunk/Projects/Framework-Samples/PLINQO/VisualBasic/Petshop/Source/Petshop.Test/DataTests.vb
Imports System
Imports System.Diagnostics
Imports NUnit.Framework
Imports PetShop.Core.Data
Imports System.Linq
Imports System.Data.Linq
'Imports PetShop.Core.Utility

Namespace PetShop.Test
    <TestFixture()> _
    Public Class BusinessTests
        <SetUp()> _
        Public Sub Setup()
        End Sub

        <TearDown()> _
        Public Sub TearDown()
        End Sub

        Private Const NAME As String = "UnitTests"
        Private Const ID As String = "Unit-Test"
        Private Shared _supplierId As Integer = 1

#Region "Profile"

        <Test()> _
        Public Sub CreateProfile()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim profile = New Profile()
            profile.Username = NAME
            profile.ApplicationName = "PetShop..Businesss"
            profile.IsAnonymous = False
            profile.LastActivityDate = DateTime.Now
            profile.LastUpdatedDate = DateTime.Now

            Try
                Using context = New PetShopDataContext()
                    context.Profile.InsertOnSubmit(profile)
                    context.SubmitChanges()
                End Using
            Catch ex As Exception
                Assert.Fail(ex.Message)
            End Try

            Assert.IsTrue(True)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub FetchProfile()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim profile As Profile = Nothing
            Using context = New PetShopDataContext()
                profile = context.Profile.ByApplicationName(NAME).First()
                profile.Detach()
            End Using

            Assert.IsTrue(profile.Username = NAME)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub UpdateProfile()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim profile As Profile = Nothing
            Using context = New PetShopDataContext()
                profile = context.Profile.ByApplicationName(NAME).First()
                profile.Detach()
            End Using

            Using context = New PetShopDataContext()
                context.Profile.Attach(profile)
                profile.IsAnonymous = True
                context.SubmitChanges()
            End Using

            Using context = New PetShopDataContext()
                Assert.IsTrue(context.Profile.ByApplicationName(NAME).First().IsAnonymous.Value)
            End Using
            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Category"

        <Test()> _
        Public Sub CreateCategory()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim category = New Category()
            category.CategoryId = ID
            category.Name = ""
            category.Descn = ""

            Try
                Using context = New PetShopDataContext()
                    context.Category.InsertOnSubmit(category)
                    context.SubmitChanges()
                End Using
            Catch ex As Exception
                Assert.Fail(ex.Message)
            End Try

            Assert.IsTrue(True)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub FetchCategory()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim category As Category = Nothing
            Using context = New PetShopDataContext()
                category = context.Category.GetByKey(ID)
                category.Detach()
            End Using
            Assert.IsTrue(category.CategoryId = ID)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub UpdateCategory()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim category As Category = Nothing
            Using context = New PetShopDataContext()
                category = context.Category.GetByKey(ID)
                category.Detach()
            End Using

            Using context = New PetShopDataContext()
                context.Category.Attach(category)
                category.Descn = "This is a ."
                context.SubmitChanges()
            End Using

            Using context = New PetShopDataContext()
                Assert.IsTrue(context.Category.GetByKey(ID).Descn = "This is a .")
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Inventory"

        <Test()> _
        Public Sub CreateInventory()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim inventory = New Inventory()
            inventory.ItemId = ID
            inventory.Qty = 10

            Try
                Using context = New PetShopDataContext()
                    context.Inventory.InsertOnSubmit(inventory)
                    context.SubmitChanges()
                End Using
            Catch ex As Exception
                Assert.Fail(ex.Message)
            End Try

            Assert.IsTrue(True)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub FetchInventory()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim inventory As Inventory = Nothing

            Using context = New PetShopDataContext()
                inventory = context.Inventory.GetByKey(ID)
                inventory.Detach()
            End Using

            Assert.IsTrue(inventory.ItemId = ID)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub UpdateInventory()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim inventory As Inventory = Nothing
            Using context = New PetShopDataContext()
                inventory = context.Inventory.GetByKey(ID)
                inventory.Detach()
            End Using

            Using context = New PetShopDataContext()
                context.Inventory.Attach(inventory)
                inventory.Qty = 100
                context.SubmitChanges()
            End Using

            Using context = New PetShopDataContext()
                Assert.IsTrue(context.Inventory.GetByKey(ID).Qty = 100)
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Product"

        <Test()> _
        Public Sub CreateProduct()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim product = New Product()
            product.ProductId = ID
            product.CategoryId = ID
            product.Image = "/.png"
            product.Descn = ""
            product.Name = ""

            Try
                Using context = New PetShopDataContext()
                    context.Product.InsertOnSubmit(product)
                    context.SubmitChanges()
                End Using
            Catch ex As Exception
                Assert.Fail(ex.Message)
            End Try

            Assert.IsTrue(True)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub FetchProduct()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim product As Product = Nothing
            Using context = New PetShopDataContext()
                product = context.Product.GetByKey(ID)
                product.Detach()
            End Using

            Assert.IsTrue(product.ProductId = ID)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub UpdateProduct()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim product As Product = Nothing
            Using context = New PetShopDataContext()
                product = context.Product.GetByKey(ID)
                product.Detach()
            End Using

            Using context = New PetShopDataContext()
                context.Product.Attach(product)
                product.Descn = "This is a "
                context.SubmitChanges()
            End Using

            Using context = New PetShopDataContext()
                Assert.IsTrue(context.Product.GetByKey(ID).Descn = "This is a ")
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Supplier"

        <Test()> _
        Public Sub CreateSupplier()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim supplier = New Supplier()
            supplier.Name = NAME
            supplier.Status = "AB"
            supplier.Addr1 = "One  Way"
            supplier.Addr2 = "Two  Way"
            supplier.City = "Dallas"
            supplier.State = "TX"
            supplier.Zip = "90210"
            supplier.Phone = "555-555-5555"

            Try
                Using context = New PetShopDataContext()
                    context.Supplier.InsertOnSubmit(supplier)
                    context.SubmitChanges()
                    supplier.Detach()
                    _supplierId = supplier.SuppId
                End Using
            Catch ex As Exception
                Assert.Fail(ex.Message)
            End Try

            Assert.IsTrue(True)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub FetchSupplier()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim supplier As Supplier = Nothing
            Using context = New PetShopDataContext()
                supplier = context.Supplier.GetByKey(_supplierId)
                supplier.Detach()
            End Using

            Assert.IsTrue(supplier.SuppId = _supplierId)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub UpdateSupplier()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim supplier As Supplier = Nothing
            Using context = New PetShopDataContext()
                supplier = context.Supplier.GetByKey(_supplierId)
                supplier.Detach()
            End Using

            Using context = New PetShopDataContext()
                context.Supplier.Attach(supplier)
                supplier.Phone = "111-111-1111"
                context.SubmitChanges()
            End Using

            Using context = New PetShopDataContext()
                Assert.IsTrue(context.Supplier.GetByKey(_supplierId).Phone = "111-111-1111")
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Item"

        <Test()> _
        Public Sub CreateItem()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim item = New Item()
            item.ItemId = ID
            item.Image = "/.png"
            item.ListPrice = 0
            item.Name = ""
            item.ProductId = ID
            item.Status = ""
            item.Supplier = _supplierId
            item.UnitCost = 0

            Try
                Using context = New PetShopDataContext()
                    context.Item.InsertOnSubmit(item)
                    context.SubmitChanges()
                End Using
            Catch ex As Exception
                Assert.Fail(ex.Message)
            End Try

            Assert.IsTrue(True)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub FetchItem()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim item As Item = Nothing
            Using context = New PetShopDataContext()
                item = context.Item.GetByKey(ID)
                item.Detach()
            End Using

            Assert.IsTrue(item.ItemId = ID)

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

        <Test()> _
        Public Sub UpdateItem()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Dim item As Item = Nothing
            Using context = New PetShopDataContext()
                item = context.Item.GetByKey(ID)
                item.Detach()
            End Using

            Using context = New PetShopDataContext()
                context.Item.Attach(item)
                item.ListPrice = 111
                context.SubmitChanges()
            End Using

            Using context = New PetShopDataContext()
                Assert.IsTrue(context.Item.GetByKey(ID).ListPrice = 111)
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Deletes"

#Region "Delete Item"

        <Test()> _
        Public Sub DeleteItem()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Using context = New PetshopDataContext()
                context.Item.Delete(ID)
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Delete Supplier"

        <Test()> _
        Public Sub DeleteSupplier()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Using context = New PetshopDataContext()
                context.Supplier.Delete(_supplierId)
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Delete Product"

        <Test()> _
        Public Sub DeleteProduct()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Using context = New PetshopDataContext()
                context.Product.Delete(ID)
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Delete Inventory"

        <Test()> _
        Public Sub DeleteInventory()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Using context = New PetshopDataContext()
                context.Inventory.Delete(ID)
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Delete Category"

        <Test()> _
        Public Sub DeleteCategory()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Using context = New PetshopDataContext()
                context.Category.Delete(ID)
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#Region "Delete Profile"

        <Test()> _
        Public Sub DeleteProfile()
            Dim watch As Stopwatch = Stopwatch.StartNew()

            Using context = New PetshopDataContext()
                Dim profile = context.Profile.ByApplicationName(NAME).First()
                context.Profile.Delete(profile.UniqueID)
            End Using

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds)
        End Sub

#End Region

#End Region
    End Class
End Namespace

File: CSharp/PetShop/PetShop.Test/DataTests.cs
using System;
using System.Diagnostics;
using NUnit.Framework;
using System.Linq;
using System.Data.Linq;
using PetShop.Core.Utility;
using Petshop.Data;
using Petshop.Data.Entities;

namespace PetShop.Test
{
    [TestFixture]
    public class BusinessTests
    {
        [SetUp]
        public void Setup()
        {
        }

        [TearDown]
        public void TearDown()
        {
        }

        private const string NAME = "UnitTests";
        private const string ID = "Unit-Test";
        private static int _supplierId = 1;

        #region Profile

        [Test]
        public void CreateProfile()
        {
            Stopwatch watch = Stopwatch.StartNew();

            var profile = new Profile();
            profile.Username = NAME;
            profile.ApplicationName = "PetShop..Businesss";
            profile.IsAnonymous = false;
            profile.LastActivityDate = DateTime.Now;
            profile.LastUpdatedDate = DateTime.Now;

            try
            {
                using (var context = new PetshopDataContext())
                {
                    context.Profile.InsertOnSubmit(profile);
                    context.SubmitChanges();
                }
            }
            catch (Exception ex)
            {
                Assert.Fail(ex.Message);
            }
            
            Assert.IsTrue(true);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void FetchProfile()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Profile profile = null;
            using (var context = new PetshopDataContext())
            {
                profile = context.Profile.GetProfile(NAME);
                context.Profile.Detach(profile);
            }

            Assert.IsTrue(profile.Username == NAME);
            
            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void UpdateProfile()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Profile profile = null;
            using (var context = new PetshopDataContext())
            {
                profile = context.Profile.GetProfile(NAME);
                context.Profile.Detach(profile);
            }

            using (var context = new PetshopDataContext())
            {
                context.Profile.Attach(profile);
                profile.IsAnonymous = true;
                context.SubmitChanges();
            }

            using (var context = new PetshopDataContext())
            {
                Assert.IsTrue(context.Profile.GetProfile(NAME).IsAnonymous.Value);
            }
            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Category

        [Test]
        public void CreateCategory()
        {
            Stopwatch watch = Stopwatch.StartNew();

            var category = new Category();
            category.CategoryId = ID;
            category.Name = "";
            category.Description = "";

            try
            {
                using (var context = new PetshopDataContext())
                {
                    context.Category.InsertOnSubmit(category);
                    context.SubmitChanges();
                }
            }
            catch (Exception ex)
            {
                Assert.Fail(ex.Message);
            }

            Assert.IsTrue(true);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void FetchCategory()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Category category = null;
            using (var context = new PetshopDataContext())
            {
                category = context.Category.GetByKey(ID);
                context.Category.Detach(category);
            }
            Assert.IsTrue(category.CategoryId == ID);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void UpdateCategory()
        {
            Stopwatch watch = Stopwatch.StartNew();
            
            Category category = null;
            using (var context = new PetshopDataContext())
            {
                category = context.Category.GetByKey(ID);
                context.Category.Detach(category);
            }

            using (var context = new PetshopDataContext())
            {
                context.Category.Attach(category);
                category.Description = "This is a .";
                context.SubmitChanges();
            }

            using (var context = new PetshopDataContext())
            {
                Assert.IsTrue(context.Category.GetByKey(ID).Description == "This is a .");
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Inventory

        [Test]
        public void CreateInventory()
        {
            Stopwatch watch = Stopwatch.StartNew();

            var inventory = new Inventory();
            inventory.ItemId = ID;
            inventory.Qty = 10;

            try
            {
                using (var context = new PetshopDataContext())
                {
                    context.Inventory.InsertOnSubmit(inventory);
                    context.SubmitChanges();
                }
            }
            catch (Exception ex)
            {
                Assert.Fail(ex.Message);
            }

            Assert.IsTrue(true);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void FetchInventory()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Inventory inventory = null;

            using (var context = new PetshopDataContext())
            {
                inventory = context.Inventory.GetByKey(ID);
                context.Inventory.Detach(inventory);
            }

            Assert.IsTrue(inventory.ItemId == ID);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void UpdateInventory()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Inventory inventory = null;
            using (var context = new PetshopDataContext())
            {
                inventory = context.Inventory.GetByKey(ID);
                context.Inventory.Detach(inventory);
            }

            using (var context = new PetshopDataContext())
            {
                context.Inventory.Attach(inventory);
                inventory.Qty = 100;
                context.SubmitChanges();
            }

            using (var context = new PetshopDataContext())
            {
                Assert.IsTrue(context.Inventory.GetByKey(ID).Qty == 100);
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Product

        /*[Test]
        public void CreateProduct()
        {
            Stopwatch watch = Stopwatch.StartNew();

            var product = new Product();
            product.ProductId = ID;
            product.CategoryId = ID;
            product.Image = "/.png";
            product.Descn = "";
            product.Name = "";

            try
            {
                using (var context = new PetshopDataContext())
                {
                    context.Product.InsertOnSubmit(product);
                    context.SubmitChanges();
                }
            }
            catch (Exception ex)
            {
                Assert.Fail(ex.Message);
            }

            Assert.IsTrue(true);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }*/

        [Test]
        public void FetchProduct()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Product product = null;
            using (var context = new PetshopDataContext())
            {
                product = context.Product.GetByKey(ID);
                context.Product.Detach(product);
            }

            Assert.IsTrue(product.ProductId == ID);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void UpdateProduct()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Product product = null;
            using (var context = new PetshopDataContext())
            {
                product = context.Product.GetByKey(ID);
                context.Product.Detach(product);
            }

            using (var context = new PetshopDataContext())
            {
                context.Product.Attach(product);
                product.Description = "This is a ";
                context.SubmitChanges();
            }

            using (var context = new PetshopDataContext())
            {
                Assert.IsTrue(context.Product.GetByKey(ID).Description == "This is a ");
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Supplier

        [Test]
        public void CreateSupplier()
        {
            Stopwatch watch = Stopwatch.StartNew();

            var supplier = new Supplier();
            supplier.Name = NAME;
            supplier.Status = "AB";
            supplier.Addr1 = "One  Way";
            supplier.Addr2 = "Two  Way";
            supplier.City = "Dallas";
            supplier.State = "TX";
            supplier.Zip = "90210";
            supplier.Phone = "555-555-5555";

            try
            {
                using (var context = new PetshopDataContext())
                {
                    context.Supplier.InsertOnSubmit(supplier);
                    context.SubmitChanges();
                    context.Supplier.Detach(supplier);
                    _supplierId = supplier.SuppId;
                }
            }
            catch (Exception ex)
            {
                Assert.Fail(ex.Message);
            }

            Assert.IsTrue(true);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void FetchSupplier()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Supplier supplier = null;
            using (var context = new PetshopDataContext())
            {
                supplier = context.Supplier.GetByKey(_supplierId);
                context.Supplier.Detach(supplier);
            }

            Assert.IsTrue(supplier.SuppId == _supplierId);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void UpdateSupplier()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Supplier supplier = null;
            using (var context = new PetshopDataContext())
            {
                supplier = context.Supplier.GetByKey(_supplierId);
                context.Supplier.Detach(supplier);
            }

            using (var context = new PetshopDataContext())
            {
                context.Supplier.Attach(supplier);
                supplier.Phone = "111-111-1111";
                context.SubmitChanges();
            }

            using (var context = new PetshopDataContext())
            {
                Assert.IsTrue(context.Supplier.GetByKey(_supplierId).Phone == "111-111-1111");
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Item

        /*[Test]
        public void CreateItem()
        {
            Stopwatch watch = Stopwatch.StartNew();

            var item = new Item();
            item.ItemId = ID;
            item.Image = "/.png";
            item.ListPrice = 0;
            item.Name = "";
            item.ProductId = ID;
            item.Status = "";
            item.Supplier1.SuppId = _supplierId;
            item.UnitCost = 0;

            try
            {
                using (var context = new PetshopDataContext())
                {
                    context.Item.InsertOnSubmit(item);
                    context.SubmitChanges();
                }
            }
            catch (Exception ex)
            {
                Assert.Fail(ex.Message);
            }

            Assert.IsTrue(true);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }*/

        [Test]
        public void FetchItem()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Item item = null;
            using (var context = new PetshopDataContext())
            {
                item = context.Item.GetByKey(ID);
                context.Item.Detach(item);
            }

            Assert.IsTrue(item.ItemId == ID);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void UpdateItem()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Item item = null;
            using (var context = new PetshopDataContext())
            {
                item = context.Item.GetByKey(ID);
                context.Item.Detach(item);
            }

            using (var context = new PetshopDataContext())
            {
                context.Item.Attach(item);
                item.ListPrice = 111;
                context.SubmitChanges();
            }

            using (var context = new PetshopDataContext())
            {
                Assert.IsTrue(context.Item.GetByKey(ID).ListPrice == 111);
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region ShoppingCart

        [Test]
        public void CreateAndFetchShoppingCart()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Profile profile = null;
            using (var context = new PetshopDataContext())
            {
                profile = context.Profile.GetProfile(NAME);
                context.Profile.Detach(profile);
            }

            Assert.IsTrue(profile.ShoppingCart.Count == 0);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void AddItemToShoppingCart()
        {
            Stopwatch watch = Stopwatch.StartNew();
            
            Profile profile = null;
            using (var context = new PetshopDataContext())
            {
                profile = context.Profile.GetProfile(NAME);
                context.Profile.Detach(profile);
            }

            using (var context = new PetshopDataContext())
            {
                CartHelper.Add(profile.ShoppingCart, ID, profile.UniqueID, true);
            }

            using (var context = new PetshopDataContext())
            {
                Assert.IsTrue(context.Profile.GetProfile(NAME).ShoppingCart.Count == 1);
            }
            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void RemoveItemFromShoppingCart()
        {
            Stopwatch watch = Stopwatch.StartNew();

            Profile profile = null;
            using (var context = new PetshopDataContext())
            {
                profile = context.Profile.GetProfile(NAME);
                context.Profile.Detach(profile);
            }

            using (var context = new PetshopDataContext())
            {
                CartHelper.Remove(profile.ShoppingCart, ID);
            }

            using (var context = new PetshopDataContext())
            {
                Assert.IsTrue(context.Profile.GetProfile(NAME).ShoppingCart.Count == 0);
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void UpdateItemQuantityShoppingCart()
        {
            Stopwatch watch = Stopwatch.StartNew();

            //Add new Item to the cart.
            Profile profile = null;
            using (var context = new PetshopDataContext())
            {
                profile = context.Profile.GetProfile(NAME);
                context.Profile.Detach(profile);
            }

            CartHelper.Add(profile.ShoppingCart, ID, profile.UniqueID, true);
            CartHelper.Add(profile.ShoppingCart, ID, profile.UniqueID, true);


            using (var context = new PetshopDataContext())
            {
                profile = context.Profile.GetProfile(NAME);
                context.Profile.Detach(profile);
            }

            Assert.IsTrue(profile.ShoppingCart.Count == 1 && profile.ShoppingCart[0].Quantity == 2);

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        [Test]
        public void ClearShoppingCart()
        {
            Stopwatch watch = Stopwatch.StartNew();

            //Add new Item to the cart.
            Profile profile = null;
            using (var context = new PetshopDataContext())
            {
                profile = context.Profile.GetProfile(NAME);
                context.Profile.Detach(profile);
            }

            CartHelper.Add(profile.ShoppingCart, ID, profile.UniqueID, true);
            

            //Clear the cart.
            using (var context = new PetshopDataContext())
            {
                profile = context.Profile.GetProfile(NAME);
                context.Profile.Detach(profile);
            }
            
            CartHelper.ClearCart(profile.ShoppingCart);

            using (var context = new PetshopDataContext())
            {
                Assert.IsTrue(context.Profile.GetProfile(NAME).ShoppingCart.Count == 0);
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Deletes

        /*#region Delete Item

        [Test]
        public void DeleteItem()
        {
            Stopwatch watch = Stopwatch.StartNew();

            using (var context = new PetshopDataContext())
            {
                context.Item.Delete(ID);
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Delete Supplier

        [Test]
        public void DeleteSupplier()
        {
            Stopwatch watch = Stopwatch.StartNew();

            using (var context = new PetshopDataContext())
            {
                context.Supplier.Delete(_supplierId);
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Delete Product

        [Test]
        public void DeleteProduct()
        {
            Stopwatch watch = Stopwatch.StartNew();

            using (var context = new PetshopDataContext())
            {
                context.Product.Delete(ID);
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Delete Inventory

        [Test]
        public void DeleteInventory()
        {
            Stopwatch watch = Stopwatch.StartNew();

            using (var context = new PetshopDataContext())
            {
                context.Inventory.Delete(ID);
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Delete Category

        [Test]
        public void DeleteCategory()
        {
            Stopwatch watch = Stopwatch.StartNew();

            using (var context = new PetshopDataContext())
            {
                context.Category.Delete(ID);
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion

        #region Delete Profile

        [Test]
        public void DeleteProfile()
        {
            Stopwatch watch = Stopwatch.StartNew();

            using (var context = new PetshopDataContext())
            {
                var profile = context.Profile.GetProfile(NAME);
                context.Profile.Delete(profile.UniqueID);
            }

            Console.WriteLine("Time: {0} ms", watch.ElapsedMilliseconds);
        }

        #endregion*/

        #endregion
    }
}
